test/e2e/storage/testsuites/volumeperf.go

```diff
@@ -0,0 +1,287 @@
+	}
+}
+
+func (t *volumePerformanceTestSuite) GetTestSuiteInfo() TestSuiteInfo {
+	return t.tsInfo
+}
+
+func (t *volumePerformanceTestSuite) SkipRedundantSuite(driver TestDriver, pattern testpatterns.TestPattern) {
+}
+
+func (t *volumePerformanceTestSuite) DefineTests(driver TestDriver, pattern testpatterns.TestPattern) {
+	type local struct {
+		config      *PerTestConfig
+		testCleanup func()
+		cs          clientset.Interface
+		ns          *v1.Namespace
+		pvcs        []*v1.PersistentVolumeClaim
+	}
+	ginkgo.BeforeEach(func() {
+		// Check preconditions
+		if dDriver = driver.(DynamicPVTestDriver); dDriver == nil {
+			framework.Failf("Test driver does not support dynamically created volumes")
+		}
+	})
+
+	f := framework.NewDefaultFramework("volume-lifecycle-performance")
+	ginkgo.It("volume lifecycle performance [Slow]", func() {
+		l := &local{
+			cs: f.ClientSet,
+			ns: f.Namespace,
+		}
+		l.config, l.testCleanup = driver.PrepareTest(f)
+		defer l.testCleanup()
+
+		// Stats for volume provisioning operation
+		// TODO: Add stats for attach, resize and snapshot?
+		provisioningStats := &performanceStats{
+			mutex:             &sync.Mutex{},
+			perObjectInterval: make(map[string]*interval),
+			operationMetrics:  &metrics{},
+		}
+
+		sc := driver.(DynamicPVTestDriver).GetDynamicProvisionStorageClass(l.config, pattern.FsType)
+		ginkgo.By(fmt.Sprintf("Creating Storage Class %s", sc.Name))
+		_, err := l.cs.StorageV1().StorageClasses().Create(context.TODO(), sc, metav1.CreateOptions{})
+		framework.ExpectNoError(err)
+		defer func() {
+			ginkgo.By(fmt.Sprintf("Deleting Storage Class %s", sc.Name))
+			err := l.cs.StorageV1().StorageClasses().Delete(context.TODO(), sc.Name, metav1.DeleteOptions{})
+			framework.ExpectNoError(err)
+		}()
+
+		// Create a controller to watch on PVCs
+		// When all PVCs provisioned by this test are in the Bound state, the controller
+		// sends a signal to the channel
+		controller := newPVCWatch(f, provisioningStats)
+		stopCh := make(chan struct{})
+		go controller.Run(stopCh)
+		defer close(stopCh)
+		waitForProvisionCh = make(chan struct{})
+
+		ginkgo.By(fmt.Sprintf("Creating %d PVCs", provisionCount))
+		for i := 0; i < provisionCount; i++ {
+			pvc := e2epv.MakePersistentVolumeClaim(e2epv.PersistentVolumeClaimConfig{
+				ClaimSize:        "5Mi",
+				StorageClassName: &sc.Name,
+			}, l.ns.Name)
+			pvc, err = l.cs.CoreV1().PersistentVolumeClaims(l.ns.Name).Create(context.TODO(), pvc, metav1.CreateOptions{})
+			framework.ExpectNoError(err)
+			// Store create time for each PVC
+			provisioningStats.mutex.Lock()
+			provisioningStats.perObjectInterval[pvc.Name] = &interval{
+				create: pvc.CreationTimestamp.Time,
+			}
+			provisioningStats.mutex.Unlock()
+			l.pvcs = append(l.pvcs, pvc)
+		}
+		defer func() {
+			ginkgo.By("Deleting all PVCs")
+			for _, pvc := range l.pvcs {
+				err := l.cs.CoreV1().PersistentVolumeClaims(l.ns.Name).Delete(context.TODO(), pvc.Name, metav1.DeleteOptions{})
```

Comment:
We also need to wait for the PV objects to be deleted. Otherwise, we may uninstall the csi driver before the pv deletion is complete

Reply:
Done